### DISCUSSION TITLE
ESP32 vs RP Pico using both cores

### DISCUSSION THREAD

**Discussion by ZacharyInDallas** [micropython/micropython#11683] [ESP32] [answered]
I’m running the same Python code on both a RP Pico and an ESP32-S3.  The Pico seems to run better in the sense that things like uasyncio.sleep returns after an appropriate amount of time (when set to a few milliseconds), but the ESP32-S3 returns after a random time period, sometime over 1 second.

From a tutorial I had watched, I though the ESP32-S3 would run the 2nd thread on the 2nd core.  After researching why the ESP32-S3 wasn’t behaving as I expected, I saw threads claiming that the ESP32-S3 would not use the 2nd core.  Is that true?

A thread mentioned that the “GIL is enabled/disabled via MICROPY_PY_THREAD_GIL”.  If you were to do that, would the ESP32-S3 then run the threads in separate processors like the Pico does, or would there be some problem?

**Comment by bixb922** [comment]
The second core of the ESP32 is used by ESP-IDF runtime and is not available for Micropython. On the other hand I do run time critical code on the ESP32S3 with asyncio and the deviations I see from the asyncio.sleep_ms time are of a few milliseconds at most (never seconds), except if some other asyncio task does not yield timely. How much RAM does the ESP32S3 have?

**Reply by ZacharyInDallas** [reply]
The board is:

ESP32-S3-DevKitC-1-N32R8V
ESP32-S3-WROOM-2-N32R8V
32 MB OT
8 MB OT
1.8 V

The other thread running on the same core is microdot:
from microdot_asyncio import Microdot, Request
from microdot_utemplate import render_template

Everything is well behaved on the Pico.  If microdot is the problem then it's not a problem on the Pico.

**Reply by bixb922** [reply, marked as answer]
Probably some asyncio task is taking that time. One thing that comes to mind is to garbage collection. SPIRAM is rather slow compared to the internal RAM of the RP2040 a ESP32, and 8Mb is probably 30 to 50 times more RAM than what's available on the RPi. Try gc.threshold and/or add a task to do garbage collection frequently, or at a times when it can't interfere.
Microdot/wifi can also take some asyncio time. To measure if there are long running tasks, I usually add another task that does a loop of asyncio.sleep(0), measuring how long that statement takes, and report the maximum value every second. 
This profiler https://github.com/micropython/micropython-lib/pull/650 could also help, but I could not find how to try it.

**Reply by peterhinch** [reply]
I ran [mico-gui](https://github.com/peterhinch/micropython-micro-gui) on an ESP32-S3 with SPIRAM. This uses `uasyncio`. Latency much beyond 100ms would be evident in a lag between pressing a button and a response on screen. It ran fine. I appreciate this isn't a very precise test, but latency of 1s would have been very obvious. So I agree with @bixb922: something is blocking.

**Reply by ZacharyInDallas** [reply]
I found the problem.  I had a second thread running, that when written, assumed it would always be the only thread running on core 2 and had no sleeps in it.  That worked fine on the Pico because it uses 2 cores.  I found out in this thread that the ESP32 only uses one of its cores for MicroPython.  On the ESP32, the 2nd thread was running on core 1 and did not have any sleeps in it, so it was hogging that core.  Adding a sleep fixed that problem.

I would say that the Pico runs my software noticeably better than the ESP32.  The Pico runs the code faster.  It is also more deterministic in that even after adding the sleep to the second thread, there is still some variation on how long the sleeps actually take.  None of this is surprising in that you are comparing threads running on 2 cores vs. threads sharing a core, even if that 1 core is faster.

**Comment by ajevticc** [comment]
Hi @ZacharyInDallas, could you please provide some sample software (or a link to) of the two threads running? I have tried with _thread module but it would only work until I would initialize pins or I2C bus.

**Reply by ZacharyInDallas** [reply]
Everything I do is from examples in the MicroPython docs.  The code below is not really different from those.  The only real issue I had was that it is not documented that on some processors your 2nd thread will run on its own core and on others it will share a core.

I use:
uasyncio.create_task (mainLoopController ())
to create my main thread.

async def mainLoopController ():
then uses:
_thread.start_new_thread (newThread, (True, True))
to execute another thread and goes into a loop:

async def newThread (
		has0: bool, 
		has1: bool):

newThread also enters a loop:
	while (True):

**Reply by ajevticc** [reply]
Ah cool, I see now, many thanks! The main tread is executed on the first core by default?

**Reply by ZacharyInDallas** [reply]
Yes, I believe so.  I think on the ESP32 you can actually have many, non async, threads running as MicroPython on the ESP32 runs on top of FreeRTOS.  On the RP Pico you are limited to 2 non async threads, 1 for each core.

**Comment by mattytrentini** [comment]
FWIW I started a wiki page on MicroPython [Threads](https://github.com/micropython/micropython/wiki/Threads) some time ago. It's incomplete but we can build on it! Let me know where you'd like additional detail and I can fill it out...

**Reply by ajevticc** [reply]
Hi @mattytrentini, thank you for your response! Here it's okay also or a private message. 

I like how you made a video about it also. https://www.youtube.com/watch?v=aDXgX0rGVDY

**Reply by mattytrentini** [reply]
Here or privately is fine. Or perhaps on the [MicroPython Discord](https://micropython.org/discord) would be best.
